ntoskrnl/ps/job.c

```diff
@@ -80,159 +84,1413 @@ ULONG PspJobInfoAlign[] =
+
+    /* Get the first process from the job */
+    Process = PspAdvanceJobEnumerator(Job, NULL);
+
+    /* Iterate through all processes in the job */
+    while (Process)
+    {
+        /* Call the provided callback */
+        Status = Callback(Process, Context);
+        if (!NT_SUCCESS(Status))
+        {
+            AnyCallbackFailed = TRUE;
+            if (BreakOnCallbackFailure)
+            {
+                break;
+            }
+        }
+
+        /* Move to the next process */
+        Process = PspAdvanceJobEnumerator(Job, Process);
+    }
+
+    if (NT_SUCCESS(Status) && AnyCallbackFailed)
+    {
+        DPRINT1("PspEnumerateProcessesInJob(Job: %p, Callback: %p, Context: %p,"
+                " BreakOnCallbackFailure: %u) - Partial success report, not all"
+                " callbacks returned success\n",
+                Job, Callback, Context, BreakOnCallbackFailure);
+    }
+
+    return Status;
+}
+
+/*!
+ * Assigns a process to a job object.
+ *
+ * @param[in] Process
+ *     Pointer to the process to be assigned to the job.
+ *
+ * @param[in] Job
+ *     Pointer to the job object to which the process is to be assigned.
+ *
+ * @returns
+ *     STATUS_SUCCESS if the process is successfully assigned to the job.
+ *     An appropriate NTSTATUS error code otherwise.
+ */
+NTSTATUS
+NTAPI
+PspAssignProcessToJob(
+    _In_ PEPROCESS Process,
+    _In_ PEJOB Job
+)
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    DPRINT1("PspAssignProcessToJob(Process: %p, Job: %p)\n", Process, Job);
+
+    ExEnterCriticalRegionAndAcquireResourceExclusive(&Job->JobLock);
+
+    /* Check if the job has a limit on the number of active processes */
+    if (Job->LimitFlags & JOB_OBJECT_LIMIT_ACTIVE_PROCESS)
+    {
+        /* Check if job limit on active processes has been reached */
+        if (Job->ActiveProcesses > Job->ActiveProcessLimit)
+        {
+            if (Job->CompletionPort)
+            {
+                /* If the job has a completion port, notify the job that the
+                   limit on the number of active processes has been exceeded */
+                Status = IoSetIoCompletion(Job->CompletionPort,
+                                           Job->CompletionKey,
+                                           NULL,
+                                           STATUS_SUCCESS,
+                                           JOB_OBJECT_MSG_ACTIVE_PROCESS_LIMIT,
+                                           TRUE);
+            }
+            else
+            {
+                Status = STATUS_QUOTA_EXCEEDED;
+            }
+
+            return Status;
+        }
+    }
+
+    /* https://learn.microsoft.com/en-us/windows/win32/api/jobapi2/nf-jobapi2-assignprocesstojobobject:
+       "If the job or any of its parent jobs in the job chain is terminating
+       when AssignProcessToJob is called, the function fails" */
+    if (Job->JobFlags & JOB_OBJECT_TERMINATING)
+    {
+        return STATUS_INVALID_PARAMETER;
```

Comment:
Leaking the lock here and below.

```suggestion
        Status = STATUS_INVALID_PARAMETER;
        goto Exit;
```